test/engine-luatest/gh_10847_memtx_sort_data_test.lua

```diff
@@ -0,0 +1,528 @@
+        local fio = require('fio')
+        local glob_snap = fio.pathjoin(box.cfg.memtx_dir, '*.snap')
+        local glob_sortdata = fio.pathjoin(box.cfg.memtx_dir, '*.sortdata')
+        local snapshots = fio.glob(glob_snap)
+        local sort_data_files = fio.glob(glob_sortdata)
+        t.assert(#snapshots > 0)
+        t.assert(#sort_data_files > 0)
+        for _, file in pairs(snapshots) do
+            fio.unlink(file)
+        end
+        for _, file in pairs(sort_data_files) do
+            fio.unlink(file)
+        end
+
+        -- Create a snapshot and sort data file.
+        box.snapshot()
+        t.assert(#fio.glob(glob_snap) == 1)
+        t.assert(#fio.glob(glob_sortdata) == 1)
+
+        -- Remove the .snap file.
+        fio.unlink(fio.glob(glob_snap)[1])
+        t.assert(#fio.glob(glob_snap) == 0)
+
+        -- Call to box.snapshot() fails: the sort data file exists already.
+        t.assert_error_msg_contains('.sortdata\' already exists', box.snapshot)
+        t.assert(#fio.glob(glob_snap) == 0)
+        t.assert(#fio.glob(glob_sortdata) == 1)
+    end)
+end
+
+-- Test that both snapshot and sort data file recreated if are absent.
+g_generic.test_remove_snap_and_sortdata = function(cg)
+    cg.server:exec(function()
+        -- Create a snapshot and sort data file.
+        box.snapshot()
+
+        -- Remove all snapshots and sort data files.
+        local fio = require('fio')
+        local glob_snap = fio.pathjoin(box.cfg.memtx_dir, '*.snap')
+        local glob_sortdata = fio.pathjoin(box.cfg.memtx_dir, '*.sortdata')
+        local snapshots = fio.glob(glob_snap)
+        local sort_data_files = fio.glob(glob_sortdata)
+        t.assert(#snapshots > 0)
+        t.assert(#sort_data_files > 0)
+        for _, file in pairs(snapshots) do
+            fio.unlink(file)
+        end
+        for _, file in pairs(sort_data_files) do
+            fio.unlink(file)
+        end
+
+        -- Call box.snapshot() once again: both files recreated.
+        box.snapshot()
+        t.assert_equals(#fio.glob(glob_snap), 1)
+        t.assert_equals(#fio.glob(glob_sortdata), 1)
+    end)
+end
+
+-- Test that .snap file creation failure is handled correctly.
+g_generic.test_snap_create_error = function(cg)
+    -- One way to fail the .snap file creation - is go out of file descriptors.
+    -- Let's try it in case the system has a sanely low ulimit.
+    local nfile
+    local ulimit = io.popen('ulimit -n')
+    if ulimit then
+        nfile = tonumber(ulimit:read())
+        ulimit:close()
+    end
+    if not nfile or nfile > 1024 then
+        -- Descriptors limit is to high, skip the test.
+        return
+    end
+
+    cg.server:exec(function()
+        -- Open files to approach the descriptors limit.
+        local fio = require('fio')
+        local files = {}
+        local count = 0
+        repeat
+            count = count + 1
+            files[count] = fio.open('/dev/null')
```

Comment:
You shouldn't assume that the limit is small. Better use error injections.

Reply:
There's no such between `xdir_create_xlog` and `memtx_sort_data_writer_create_file`. Do you mean to add one?